### ISSUE TITLE
Fortran dies with internal compiler error with complex array indexing

### ISSUE THREAD

**Issue by sriram-LANL** [iains/gcc-13-branch#16]
- gcc-13.2.0 installed via home-brew
- Dies on Mac OS Ventura and Sonoma
- Dies on Apple M1 Max and Apple M2 Air
- ~~Simple workaround enabled with `-DLIVE_ON` at the command line~~
- Code works fine on Linux / gcc-13

**I apologize for the confusion, simpler example in next post**
```Fortran
Derived type example deleted in favor of simpler one in next post  
```

compiler version: 
```shell
% gfortran-13 --version
GNU Fortran (Homebrew GCC 13.2.0) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by sriram-LANL** [comment]
Simpler example:

```Fortran
module reaper
  public
  real, allocatable :: xblue(:)
  real, allocatable :: xpink(:)
  integer, allocatable  :: xgreen(:)

contains
  subroutine baz()

    ASSOCIATE ( &
         blue => xblue, &
         pink => xpink, &
         green => xgreen &
         )

      blue(1) = pink(green(1))
      blue = blue/20.0

    END ASSOCIATE
  end subroutine baz
end module reaper
```

```shell
% gfortran-13 -c test.F90
f951: internal compiler error: Segmentation fault: 11
Please submit a full bug report, with preprocessed source (by using -freport-bug).
See <https://github.com/Homebrew/homebrew-core/issues> for instructions.
```

**Comment by iains** [comment]
thanks for the report

I can confirm this (and it appears to be a regression from GCC-12, but also fixed on trunk - which will become GCC-14).  At present, I need to ascertain if this is a problem with the macOS branch (in which case it's for us to fix) or whether the problem also exists for "upstream" - in which case I need to report it in GCC's Bugzilla.

This does not seem to repeat for x86_64, so would appear to be a target problem rather than a Fortran front end one.

@fxcoudert FIO (unless you happen to recognise the issue and know the solution)

**Comment by fxcoudert** [comment]
This is https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111853
I can confirm it is fixed in latest version (which will become GCC 14).

**Comment by iains** [comment]
Hmm. I hope a back port will be allowed, if not, then I will pick it onto the Darwin branch (it's a bit unfortunate that it does not show reliably - e.g. it does not fire for me with the test-case here on x86_64).

**Comment by sriram-LANL** [comment]
On godbolt, it seems to generate failures for all flavors of 13.2.  And it appears to be specific to 13.2 - trunk and 13.1 on all architectures are able to compile it.

~~Given that it is not Mac specific, @iains  would you still be willing to file a bug report with gcc?  It's hard for me to get compilers on our machines short of compiling them myself.~~

**Comment by fxcoudert** [comment]
@sriram-LANL I do not understand your comment. The bug is known, it is already reported to GCC at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111853 and it is fixed. But the fix is not in a released version yet.

**Comment by iains** [comment]
In the short-term, if you are able to self-build, I can confirm that the patch from the PR mentioned above does fix the issue on 13.2 (I would be hopeful of a backport for 13.3 - given that the bug was reported against 13.2)

```
diff --git a/gcc/fortran/resolve.cc b/gcc/fortran/resolve.cc
index 417b20637e3..c81620b9f95 100644
--- a/gcc/fortran/resolve.cc
+++ b/gcc/fortran/resolve.cc
@@ -5638,7 +5638,7 @@ gfc_expression_rank (gfc_expr *e)
       if (ref->type != REF_ARRAY)
        continue;
 
-      if (ref->u.ar.type == AR_FULL)
+      if (ref->u.ar.type == AR_FULL && ref->u.ar.as)
        {
          rank = ref->u.ar.as->rank;
          break;
```

**Comment by sriram-LANL** [comment]
> @sriram-LANL I do not understand your comment. The bug is known, it is already reported to GCC at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111853 and it is fixed. But the fix is not in a released version yet.

@fxcoudert you're right - it was the end of a long day...

**Comment by iains** [comment]
I believe this is fixed in 13.3-darwin-r0
